internal/components/generic_parser.go

```diff
@@ -30,12 +30,29 @@ var (
 // GenericParser serves as scaffolding for custom parsing logic by isolating
 // functionality to idempotent functions.
 type GenericParser[T any] struct {
-	name         string
-	settings     *Settings[T]
-	portParser   PortParser[T]
-	rbacGen      RBACRuleGenerator[T]
-	livenessGen  ProbeGenerator[T]
-	readinessGen ProbeGenerator[T]
+	name            string
+	settings        *Settings[T]
+	portParser      PortParser[T]
+	rbacGen         RBACRuleGenerator[T]
+	livenessGen     ProbeGenerator[T]
+	readinessGen    ProbeGenerator[T]
+	defaultsApplier Defaulter[T]
+}
+
+func (g *GenericParser[T]) GetDefaultConfig(logger logr.Logger, config interface{}) (interface{}, error) {
+	if g.settings == nil || g.defaultsApplier == nil {
+		return config, nil
+	}
+
+	if g.settings.defaultRecAddr == "" {
+		return config, nil
+	}
+
+	var parsed T
+	if err := mapstructure.Decode(config, &parsed); err != nil {
+		return nil, err
+	}
+	return g.defaultsApplier(logger, func(string) (string, int32) { return g.settings.defaultRecAddr, g.settings.GetServicePort().Port }, parsed)
```

Comment:
this signature feels a bit odd. The multi endpoint struct doesn't use the generic parser at all, so i think we could get away with the signature just taking in a defaultRecAddr and a port int. The multi endpoint would just use the protocol it's been given..\

Reply:
I see the point. Earlier it was that way, but I tried to reuse the same `Defaulter` on the `GenericParser`  and `MultiPortReceiver`. While we would change the signature to this:
```diff
- type Defaulter[ComponentConfigType any] func(logger logr.Logger, addrProv AddressProvider, config ComponentConfigType) (map[string]interface{}, error)
+ type Defaulter[ComponentConfigType any] func(logger logr.Logger, addrProv AddressProvider, config *SingleEndpointConfig) (map[string]interface{}, error)
 ```
Its sufficent for the `AddressDefaulter` but not for the `MultiAddressDefaulter`. Since there we may have different  addresses set and/or different default ports.

The only thing that comes to my would be moving some of the logic into the `MultiPortReceiver) GetDefaultConfig` and use the  same AddressDefaulter.